daemon/emer-aggregate-tally.c

```diff
@@ -0,0 +1,391 @@
+ * option) any later version.
+ *
+ * eos-event-recorder-daemon is distributed in the hope that it will be
+ * useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+ * Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with eos-event-recorder-daemon.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+#include "emer-aggregate-tally.h"
+
+#include <gio/gio.h>
+
+#define AGGREGATE_TIMER_VARIANT_TYPE "(vvuumv)"
+
+struct _EmerAggregateTally
+{
+  GObject parent_instance;
+
+  gchar *persistent_cache_directory;
+};
+
+G_DEFINE_TYPE (EmerAggregateTally, emer_aggregate_tally, G_TYPE_OBJECT)
+
+enum {
+  PROP_0,
+  PROP_PERSISTENT_CACHE_DIRECTORY,
+  N_PROPS,
+};
+
+static GParamSpec *properties[N_PROPS] = { NULL, };
+
+static void
+ensure_folder_for_date (EmerAggregateTally  *self,
+                        const char          *date,
+                        GError             **error)
+{
+  g_autoptr(GError) local_error = NULL;
+  g_autoptr(GFile) file = NULL;
+  g_autofree gchar *path = NULL;
+
+  path = g_build_filename (self->persistent_cache_directory,
+                           "aggregate-timers",
+                           date,
+                           NULL);
+  file = g_file_new_for_path (path);
+  g_file_make_directory_with_parents (file, NULL, &local_error);
+
+  if (local_error)
+    {
+      if (g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_EXISTS))
+        return;
+      g_propagate_error (error, g_steal_pointer (&local_error));
+    }
+}
+
+static inline GVariant *
+aggregate_timer_to_variant (guint32    unix_user_id,
+                            GVariant  *event_id,
+                            GVariant  *aggregate_key,
+                            GVariant  *payload,
+                            guint32    counter,
+                            gint64     monotonic_time_us)
+{
+  return g_variant_new (AGGREGATE_TIMER_VARIANT_TYPE,
+                        event_id,
+                        aggregate_key,
+                        unix_user_id,
+                        counter,
+                        payload);
+}
+
+static gchar *
+get_tally_path_from_aggregate_timer (EmerAggregateTally *self,
+                                     guint32             unix_user_id,
+                                     GVariant           *event_id,
+                                     GVariant           *aggregate_key,
+                                     const char         *date)
+{
+  g_autofree gchar *aggregate_key_str = NULL;
+  g_autofree gchar *event_id_str = NULL;
+  g_autofree gchar *filename = NULL;
+
+  aggregate_key_str = g_variant_print (aggregate_key, FALSE);
+  event_id_str = g_variant_print (event_id, FALSE);
+  filename = g_strdup_printf ("%u-%u-%u",
+                              g_str_hash (aggregate_key_str),
```

Comment:
(I don’t know the full context surrounding this, but…)

In general it’s better (imo) to use escaping, rather than hashing, because that allows better debuggability. It’s scary to try and debug a directory full of files which are named after checksums. Less scary to debug a directory full of files with human-readable-but-escaped names.

Reply:
I've replaced it with a SHA256sum of all the fields we use to hash each timer.